[BUG]: Unable to Add Lambda with `noexcept`

### Required prerequisites

- [X] Make sure you've read the [documentation](https://pybind11.readthedocs.io). Your issue may be addressed there.
- [X] Search the [issue tracker](https://github.com/pybind/pybind11/issues) and [Discussions](https:/pybind/pybind11/discussions) to verify that this hasn't already been reported. +1 or comment there if it has.
- [X] Consider asking first in the [Gitter chat room](https://gitter.im/pybind/Lobby) or in a [Discussion](https:/pybind/pybind11/discussions/new).

### What version (or hash if on master) of pybind11 are you using?

2.10.3

### Problem description

When trying to add lambda with noexcept, compiling fails.
Begin from the code in [documentation](https://pybind11.readthedocs.io/en/stable/advanced/exceptions.html#unraisable-exceptions)
```
#include <pybind11/pybind11.h>

namespace py = pybind11;

int nonthrowing_func() noexcept {
	try
	{
	}
	catch (py::error_already_set &eas)
	{
		eas.discard_as_unraisable(__func__);
	}
	catch (const std::exception &e)
	{
	}
	return 0;
}

PYBIND11_MODULE(m, m)
{
	m.def("a", nonthrowing_func);
}
```
this works fine.
However, if the function is turned into lambda,
```
#include <pybind11/pybind11.h>

namespace py = pybind11;

PYBIND11_MODULE(m, m)
{
	m.def("b", [](void) noexcept(true) {
		try
		{
		}
		catch (py::error_already_set &eas)
		{
			eas.discard_as_unraisable(__func__);
		}
		catch(const std::exception& e)
		{
		}
		return 0;
	});
}
```
the compiler would complain as following:
```
[1/2] Building CXX object CMakeFiles/m.dir/export.cpp.o
FAILED: CMakeFiles/m.dir/export.cpp.o 
/usr/bin/c++ -Dm_EXPORTS -isystem /usr/include/python3.10 -fPIC -fvisibility=hidden -flto -fno-fat-lto-objects -MD -MT CMakeFiles/m.dir/export.cpp.o -MF CMakeFiles/m.dir/export.cpp.o.d -o CMakeFiles/m.dir/export.cpp.o -c /home/kaigu/code/exact_solution/test_lambda/export.cpp
In file included from /usr/include/pybind11/attr.h:13,
                 from /usr/include/pybind11/detail/class.h:12,
                 from /usr/include/pybind11/pybind11.h:13,
                 from /home/kaigu/code/exact_solution/test_lambda/export.cpp:1:
/usr/include/pybind11/detail/common.h: In instantiation of ‘struct pybind11::detail::strip_function_object<pybind11_init_m(pybind11::module_&)::<lambda()> >’:
/usr/include/pybind11/detail/common.h:872:7:   required by substitution of ‘template<class Function, class F> using function_signature_t = std::conditional_t<std::is_function<_NestedType>::value, F, typename std::conditional<(std::is_pointer<_T2>::value || std::is_member_pointer<F>::value), std::remove_pointer<F>, pybind11::detail::strip_function_object<F> >::type::type> [with Function = pybind11_init_m(pybind11::module_&)::<lambda()>; F = pybind11_init_m(pybind11::module_&)::<lambda()>]’
/usr/include/pybind11/pybind11.h:101:36:   required from ‘pybind11::cpp_function::cpp_function(Func&&, const Extra& ...) [with Func = pybind11_init_m(pybind11::module_&)::<lambda()>; Extra = {pybind11::name, pybind11::scope, pybind11::sibling}; <template-parameter-1-3> = void]’
/usr/include/pybind11/pybind11.h:1153:22:   required from ‘pybind11::module_& pybind11::module_::def(const char*, Func&&, const Extra& ...) [with Func = pybind11_init_m(pybind11::module_&)::<lambda()>; Extra = {}]’
/home/kaigu/code/exact_solution/test_lambda/export.cpp:7:7:   required from here
/usr/include/pybind11/detail/common.h:867:11: error: no type named ‘type’ in ‘struct pybind11::detail::remove_class<int (pybind11_init_m(pybind11::module_&)::<lambda()>::*)() const noexcept>’
  867 |     using type = typename remove_class<decltype(&F::operator())>::type;
      |           ^~~~
ninja: build stopped: subcommand failed.
```
Changing `noexcept(true)` to `noexcept` does not help, but removing it would fix the issue.

The `CMakeLists.txt` is simple too:
```
cmake_minimum_required(VERSION 3.4)
project(example LANGUAGES CXX)

find_package(pybind11 REQUIRED CONFIG)
message(STATUS "Found pybind11 version ${pybind11_VERSION}")

pybind11_add_module(m "export.cpp")
```

### Reproducible example code

```text
See above
```


### Is this a regression? Put the last known working version here if it is.

Not a regression

--- a/include/pybind11/detail/common.h
+++ b/include/pybind11/detail/common.h
@@ -754,7 +754,16 @@ template <typename C, typename R, typename... A>
 struct remove_class<R (C::*)(A...) const> {
     using type = R(A...);
 };
-
+#ifdef __cpp_noexcept_function_type
+template <typename C, typename R, typename... A>
+struct remove_class<R (C::*)(A...) noexcept> {
+    using type = R(A...);
+};
+template <typename C, typename R, typename... A>
+struct remove_class<R (C::*)(A...) const noexcept> {
+    using type = R(A...);
+};
+#endif
 /// Helper template to strip away type modifiers
 template <typename T>
 struct intrinsic_type {

--- a/tests/test_constants_and_functions.cpp
+++ b/tests/test_constants_and_functions.cpp
@@ -149,6 +149,6 @@ TEST_SUBMODULE(constants_and_functions, m) {
             py::arg_v("z", default_value));
     });
 
-    // test noexcept(true) lambda
+    // test noexcept(true) lambda (#4565)
     m.def("l1", []() noexcept(true) { return 0; });
 }

--- a/tests/test_constants_and_functions.cpp
+++ b/tests/test_constants_and_functions.cpp
@@ -150,12 +150,5 @@ TEST_SUBMODULE(constants_and_functions, m) {
     });
 
     // test noexcept(true) lambda
-    m.def("l1", [](void) noexcept(true) {
-        try {
-        } catch (py::error_already_set &eas) {
-            eas.discard_as_unraisable(__func__);
-        } catch (const std::exception &e) {
-        }
-        return 0;
-    });
+    m.def("l1", [](void) noexcept(true) { return 0; });
 }

--- a/tests/test_constants_and_functions.cpp
+++ b/tests/test_constants_and_functions.cpp
@@ -150,5 +150,5 @@ TEST_SUBMODULE(constants_and_functions, m) {
     });
 
     // test noexcept(true) lambda
-    m.def("l1", [](void) noexcept(true) { return 0; });
+    m.def("l1", []() noexcept(true) { return 0; });
 }

--- a/tests/test_constants_and_functions.cpp
+++ b/tests/test_constants_and_functions.cpp
@@ -148,4 +148,14 @@ TEST_SUBMODULE(constants_and_functions, m) {
             py::arg_v("y", 42, "<the answer>"),
             py::arg_v("z", default_value));
     });
+
+    // test noexcept(true) lambda
+    m.def("l1", [](void) noexcept(true) {
+        try {
+        } catch (py::error_already_set &eas) {
+            eas.discard_as_unraisable(__func__);
+        } catch (const std::exception &e) {
+        }
+        return 0;
+    });
 }

--- a/tests/test_constants_and_functions.py
+++ b/tests/test_constants_and_functions.py
@@ -50,3 +50,7 @@ def test_function_record_leaks():
         m.register_large_capture_with_invalid_arguments(m)
     with pytest.raises(RuntimeError):
         m.register_with_raising_repr(m, RaisingRepr())
+
+
+def test_noexcept_lambda():
+    assert m.l1() == 0

